README.md
qgodville
=========

debian/changelog
qgodville (1.0.3) UNRELEASED; urgency=low
 

 -- binque <binque@binque-desktop>  Wed, 02 Jan 2013 12:09:41 +0400

qgodville (1.0.1) precise; urgency=low

  * 

 -- binque <binque@binque-desktop>  Wed, 17 Oct 2012 21:42:08 +0400

qgodville (1.0.0) unstable; urgency=low
  * <Add change description here>

 -- binque <binque@ya.ru>  Fri, 11 Oct 2012 19:46:51 +0400

qgodville (0.9.1) unstable; urgency=low
  * <Add change description here>

 -- binque <binque@ya.ru>  Fri, 12 Oct 2012 22:46:50 +0400

qgodville (0.0.1) unstable; urgency=low

  * Initial Release.

 -- binque <binque@ya.ru>  Fri, 12 Oct 2012 22:42:53 +0400


Godville.h
#ifndef GODVILLE_H
#define GODVILLE_H

#include <QtGui>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QNetworkAccessManager>
#ifdef Q_WS_MAEMO_5
#include <QtMaemo5/QMaemo5InformationBox>
#endif
#include "settingsdialog.h"
#include "parser.h"

class QLabel;

class Godville: public QMainWindow
{
    Q_OBJECT
public:
    Godville(QMainWindow *parent=0);
    ~Godville();

public slots:
    void updateData();
    void reloadJson();
    void readJson(QNetworkReply* list);
    void forcedUpdate();
    void options();
    void about();
    void changeFont();
    void canceled();

private:
    QLabel *nameLabel;     //hero
    QLabel *genderLabel;
    QLabel *mottoLabel;
    QLabel *alignmentLabel;
    QLabel *clanLabel;
    QLabel *levelLabel;       //data
    QLabel *inventoryLabel;
    QLabel *healthLabel;
    QLabel *questLabel;
    QLabel *goldLabel;
    QLabel *bcLabel;
    QLabel *bricksLabel;
    QLabel *diLabel;
    QLabel *distanceLabel;
    QLabel *petNameLabel;  //pet
    QLabel *petClassLabel;
    QLabel *petLevelLabel;
    QLabel *diaryLabel;       //diary
    QLabel *energyLabel;   //prana
    QProgressBar *experienceBar;
    QProgressBar *inventoryBar;
    QProgressBar *healthBar;
    QProgressBar *questBar;
    QProgressBar *energyBar;
    QGroupBox *energyBox;
    QGroupBox *petBox;
    QGroupBox *diaryBox;
    QGroupBox *inventoryBox;
    QVBoxLayout *inventoryVLayout;
    QList<QLabel*> inventoryList;
    QNetworkAccessManager *manager;
    QSettings settings;
    QTimer *timer;
    QMenu *menu;
    QAction *optionsAction;
    QAction *updateAction;
    QAction *aboutAction;
    QAction *quitAction;
    bool expired;
    bool fight;
    QByteArray json;
    bool ok;
    inline void setBarStyle (QProgressBar *bar, QString color);
};
#endif


parser.cpp
#include "parser.h"

Parser::Parser(QByteArray ba, QObject * parent) :
    QObject(parent), json(ba)
{
    parsed = new QVariantMap();
}

bool Parser::FSM()
{
    enum States { start=0, ob=1, name=2, eofName=3, colon=4, strVal=5, eofVal=6, numVal=7, comma=8, cb=9,
                  t=10, r=11, u=12, f=13, a=14, l=15, s=16, n=17, ull=18, ll=19, err=101, OK=102};
    States ST[20][17] =
    {
//       {       "       :       ,       0..9    rest    }       \0     t       r       u       e       f       a       l       s       n
    {   ob,      err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //start
    {   err,     name,   err,    err,    err,    err,    cb,     err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //ob
    {   name,    eofName,name,   name,   name,   name,   name,   err,   name,   name,   name,   name,   name,   name,   name,   name,   name}, //name
    {   err,     err,    colon,  err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //eofName
    {   ob,      strVal, err,    err,    numVal, err,    err,    err,   t,      err,    err,    err,    f,      err,    err,    err,    n}, //colon
    {   strVal,  eofVal, strVal, strVal, strVal, strVal, strVal, err,   strVal, strVal, strVal, strVal, strVal, strVal, strVal, strVal, strVal}, //strVal
    {   err,     err,    err,    comma,  err,    err,    cb,     err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //eofVal
    {   err,     err,    err,    comma,  numVal, err,    cb,     err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //numVal
    {   err,     name,   err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    err,    err,    err}, //comma
    {   err,     err,    err,    comma,  err,    err,     cb,     OK,   err,    err,    err,    err,    err,    err,    err,    err,    err},  //cb
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    r,      err,    err,    err,    err,    err,    err,    err}, //t
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    u,      err,    err,    err,    err,    err,    err}, //r
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    eofVal, err,    err,    err,    err,    err}, //u
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    a,      err,    err,    err}, //f
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    l,      err,    err}, //a
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    err,    s,      err}, //l
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    eofVal, err,    err,    err,    err,    err}, //s
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    ull,    err,    err,    err,    err,    err,    err}, //n
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    ll,     err,    err}, //u
    {   err,     err,    err,    err,    err,    err,    err,    err,   err,    err,    err,    err,    err,    err,    eofVal, err,    err}  //l
    };
    States state = start;
    int brace = 0;
    int i;
    for (i = 0; i<json.size(); i++)
    {
        switch (json.at(i))
        {
        case '{':
            state = ST[state][0];
            brace++;
            break;
        case '"':
            state = ST[state][1];
            break;
        case ':':
            state = ST[state][2];
            break;
        case ',':
            state = ST[state][3];
            break;
        case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7': case '8': case '9':
            state = ST[state][4];
            break;
        case '}':
            state = ST[state][6];
            brace--;
            break;
        case 't':
            state = ST[state][8];
            break;
        case 'r':
            state = ST[state][9];
            break;
        case 'u':
            state = ST[state][10];
            break;
        case 'e':
            state = ST[state][11];
            break;
        case 'f':
            state = ST[state][12];
            break;
        case 'a':
            state = ST[state][13];
            break;
        case 'l':
            state = ST[state][14];
            break;
        case 's':
            state = ST[state][15];
            break;
        case 'n':
            state = ST[state][16];
            break;
        default:
            state = ST[state][5];
            break;
        }
        if (state==err or state==OK)
            break;
    }
    if (state == err or brace)
    {
//        qDebug("Number of braces is %d", brace);
//        qDebug("State is %d", state);
//        qDebug("Last symbols are %c, %c, %c", json.at(i-2), json.at(i-1), json.at(i));
//        qDebug("i=%d", i);
        return false;
    }
    state = ST[state][7];
    return (state==OK);
}

bool Parser::parse()
{
    if (!FSM())
    {
        qDebug("fsm returns false");
        return false;
    }
    int i = 1;
    parsing(parsed, i);
    return true;
}

void Parser::parsing(QVariantMap *map, int &i)
{
    QString name;
    int intValue = 0;
    QString strValue;
    QVariantMap* mapValue;
    while (i<json.size())           // "
    {
        i++;                        // name
        while (json.at(i)!='"')
        {
            name += json.at(i);
            i++;
        }                           // "
        convert(&name);
        i++;                        // :
        i++;                        // " or bool or number or { or null
        switch (json.at(i))
        {
        case '"':
            i++;                    // value or "
            while (json.at(i)!='"')
            {
                strValue += json.at(i);
                i++;
            }                       // "
            i++;                    // , or }
            convert(&strValue);
            map->insert(name, QVariant(strValue));
            strValue.clear();
            name.clear();
            break;
        case 'f':
            i += 5;                 // , or }
            map->insert(name, QVariant(false));
            name.clear();
            break;
        case 't':
            i += 4;                 // , or }
            map->insert(name, QVariant(true));
            name.clear();
            break;
        case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7': case '8': case '9':
            while (json.at(i)>='0' and json.at(i)<='9')
            {
                intValue *= 10;
                intValue += (int)(json.at(i)-'0');
                i++;
            }                       // , or }
            map->insert(name, QVariant(intValue));
            intValue = 0;
            name.clear();
            break;
        case 'n':
            i += 4;                 // , or }
            name.clear();
            break;
        case '{':
            i++;                    // " or }
            mapValue = new QVariantMap;
            if (json.at(i)!='}')
                parsing(mapValue, i);// , or }
            else
                i++;                // , or }
            map->insert(name, QVariant(*mapValue));
            name.clear();
            break;
        }
        switch (json.at(i))
        {
        case ',':
            i++;                    // "
            break;
        case '}':
            i++;                    // , or }
            return;
            break;
        }
    }
}

void Parser::convert(QString *string)
{
    int i = -1;
    while ((i = string->indexOf("\\u")) != -1)
    {
        int nHex = string->mid(i + 2, 4).toInt(0, 16);
        string->replace(i, 6, QChar(nHex));
    }
}

QVariantMap *Parser::toMap()
{
    return parsed;
}

Parser::~Parser()
{
    delete parsed;
}


settingsdialog.cpp
#include "settingsdialog.h"

SettingsDialog::SettingsDialog(QWidget *parent):
    QDialog(parent)
{
    QLabel *nameLabel = new QLabel(tr("Имя бога"));
    nameLine = new QLineEdit;
    if (settings.contains("godname"))
        nameLine->setText(settings.value("godname").toString());
    QLabel *serverLabel = new QLabel(tr("Сервер"));
    serverBox = new QComboBox;
    serverBox->setEditable(false);
    serverBox->addItem("godville.net");
    serverBox->addItem("godvillegame.com");
    if (settings.value("server").toString()=="godvillegame.com")
        serverBox->setCurrentIndex(1);
    QLabel *fontLabel = new QLabel(tr("Шрифт"));
    fontBox = new QFontComboBox;
    QFont currentFont(settings.value("fontFamily").toString());
    fontBox->setCurrentFont(currentFont);
    fontSizeBox = new QComboBox;
    fontSizeBox->setEditable(false);
    for (int i=8; i<30; i+=2)
        fontSizeBox->addItem(QString::number(i));
    fontSizeBox->setCurrentIndex((settings.value("fontSize").toInt()-8)/2);
    ok = new QPushButton(tr("Применить"));
    QGridLayout *layout = new QGridLayout;
    layout->addWidget(nameLabel, 0, 0);
    layout->addWidget(nameLine, 0, 1, 1, 2);
    layout->addWidget(serverLabel, 1, 0);
    layout->addWidget(serverBox, 1, 1, 1, 2);
    layout->addWidget(fontLabel, 2, 0);
    layout->addWidget(fontBox, 2, 1);
    layout->addWidget(fontSizeBox, 2, 2);
    QVBoxLayout *formLayout = new QVBoxLayout;
    formLayout->addLayout(layout);
#ifdef Q_WS_MAEMO_5
    formLayout->addWidget(ok);
#else
    cancel = new QPushButton(tr("Отмена"));
    QHBoxLayout *buttonLayout = new QHBoxLayout;
    buttonLayout->addStretch();
    buttonLayout->addWidget(ok);
    buttonLayout->addWidget(cancel);
    formLayout->addLayout(buttonLayout);
    connect(cancel, SIGNAL(clicked()), this, SLOT(close()));
#endif
    setLayout(formLayout);

    connect(ok, SIGNAL(clicked()), this, SLOT(save()));
}

void SettingsDialog::save()
{
    settings.setValue("godname", nameLine->text());
    settings.setValue("fontFamily", fontBox->currentFont().family());
    settings.setValue("fontSize", fontSizeBox->currentText());
    if ((settings.contains("server") and settings.value("server").toString()!=serverBox->currentText())
            or (!settings.contains("server") and ((serverBox->currentText()=="godvillegame.com" and QLocale::system().name()=="ru_RU")
                                                  or (serverBox->currentText()=="godville.net" and QLocale::system().name()!="ru_RU"))))
    {
        settings.setValue("server", serverBox->currentText());
        settings.sync();
        QMessageBox::information(this, "The language changed.",
          "Please restart the program.");
        QApplication::quit();
    }
    else
        settings.setValue("server", serverBox->currentText());
    settings.sync();
    accept();
}


qgodville_en.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.0" language="en_US">
<defaultcodec>UTF-8</defaultcodec>
<context>
    <name>Godville</name>
    <message>
        <location filename="Godville.cpp" line="6"/>
        <location filename="Godville.cpp" line="20"/>
        <source>Имя</source>
        <translation>Name</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="7"/>
        <source>Пол</source>
        <translation>Gender</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="8"/>
        <source>Девиз</source>
        <translation>Motto</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="9"/>
        <source>Характер</source>
        <translation>Personality</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="10"/>
        <source>Гильдия</source>
        <translation>Guild</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="11"/>
        <location filename="Godville.cpp" line="19"/>
        <source>Уровень</source>
        <translation>Level</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="12"/>
        <source>Инвентарь</source>
        <translation>Inventory</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="13"/>
        <source>Здоровье</source>
        <translation>Health</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="14"/>
        <source>Задание</source>
        <translation>Quest</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="15"/>
        <source>Золота</source>
        <translation>Gold</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="16"/>
        <location filename="Godville.cpp" line="377"/>
        <source>Кирпичей для храма</source>
        <translation>Bricks for Temple</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="17"/>
        <location filename="Godville.cpp" line="298"/>
        <source>Столбов от столицы</source>
        <translation>Milestones Passed</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="18"/>
        <source>Вид</source>
        <translation>Class</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="21"/>
        <source>Прана</source>
        <translation>Godpower</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="155"/>
        <source>Герой</source>
        <translation>Hero</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="157"/>
        <source>Данные героя</source>
        <translation>Hero&apos;s data</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="159"/>
        <source>Питомец</source>
        <translation>Pet</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="161"/>
        <source>Дневник героя</source>
        <translation>Hero&apos;s Diary</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="163"/>
        <source>Вмешательство</source>
        <translation>Remote Control</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="165"/>
        <location filename="Godville.cpp" line="344"/>
        <source>В карманах можно найти:</source>
        <translation>Inventory</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="192"/>
        <source>Обновить</source>
        <translation>Update</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="193"/>
        <source>Настройки</source>
        <translation>Options</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="194"/>
        <location filename="Godville.cpp" line="427"/>
        <source>О программе</source>
        <translation>About</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="195"/>
        <source>Выход</source>
        <translation>Exit</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="240"/>
        <source>Ошибка загрузки данных</source>
        <translation>Error loading data</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="241"/>
        <source>Проверьте соединение с интернетом (код ошибки %1)</source>
        <translation>Please check the connection to Internet (error code %1)</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="255"/>
        <source>Ошибка распаковки данных</source>
        <translation>Error unpacking data</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="256"/>
        <source>Проверьте имя бога и сервер</source>
        <translation>Please check god&apos;s name and server</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="262"/>
        <source>мужской</source>
        <translation>male</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="264"/>
        <source>женский</source>
        <translation>female</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="268"/>
        <source>%1 (%2)</source>
        <translation>%1 (%2)</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="271"/>
        <source>(не состоит)</source>
        <translation>(no guild)</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="275"/>
        <location filename="Godville.cpp" line="278"/>
        <source>%1/%2</source>
        <translation>%1/%2</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="299"/>
        <source>%1 шт.</source>
        <translation>%1</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="304"/>
        <source>Город</source>
        <translation>Town</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="326"/>
        <source>%1 (%2 шт)</source>
        <translation>%1 (%2 pcs)</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="331"/>
        <source>%1 (@)</source>
        <translation>%1 (@)</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="342"/>
        <source>В карманах пусто.</source>
        <translation>Inventory is empty.</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="402"/>
        <source>Данные устарели!
Откройте страницу героя через браузер или полноценный клиент.</source>
        <translation>Data are outdated
Open the hero through a browser or a full-fledged client.</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="405"/>
        <location filename="Godville.cpp" line="412"/>
        <source>Битва началась!</source>
        <translation>The battle has begun!</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="409"/>
        <source>Данные устарели!</source>
        <translation>Data are outdated!</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="410"/>
        <source>Откройте страницу героя через браузер или полноценный клиент.</source>
        <translation>Open the hero through a browser or a full-fledged client.</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="412"/>
        <source>Внимание!</source>
        <translation>Warning!</translation>
    </message>
    <message>
        <location filename="Godville.cpp" line="428"/>
        <source>QGodville 1.0 
binque (Семикрылый), 2012</source>
        <translation>QGodville 1.0
binque, 2012</translation>
    </message>
    <message>
        <source>О программе QGodville</source>
        <translation type="obsolete">About QGodville</translation>
    </message>
    <message>
        <source>Разработчик: binque (Семикрылый)</source>
        <translation type="obsolete">Developer by binque</translation>
    </message>
</context>
<context>
    <name>SettingsDialog</name>
    <message>
        <location filename="settingsdialog.cpp" line="6"/>
        <source>Имя бога</source>
        <translation>God&apos;s name</translation>
    </message>
    <message>
        <location filename="settingsdialog.cpp" line="10"/>
        <source>Сервер</source>
        <translation>Server</translation>
    </message>
    <message>
        <location filename="settingsdialog.cpp" line="17"/>
        <source>Шрифт</source>
        <translation>Font</translation>
    </message>
    <message>
        <location filename="settingsdialog.cpp" line="26"/>
        <source>Применить</source>
        <translation>Apply</translation>
    </message>
    <message>
        <location filename="settingsdialog.cpp" line="40"/>
        <source>Отмена</source>
        <translation>Cancel</translation>
    </message>
</context>
</TS>
